pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,232 @@
+package restrictusers
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	usercache "github.com/openshift/origin/pkg/user/cache"
+)
+
+func init() {
+	admission.RegisterPlugin("openshift.io/RestrictSubjectBindings",
+		func(kclient kclientset.Interface, config io.Reader) (admission.Interface,
+			error) {
+			return NewRestrictUsersAdmission(kclient)
+		})
+}
+
+// restrictUsersAdmission implements admission.Interface and enforces
+// restrictions on adding rolebindings in a project to permit only designated
+// subjects.
+type restrictUsersAdmission struct {
+	*admission.Handler
+
+	oclient    oclient.Interface
+	kclient    kclientset.Interface
+	groupCache *usercache.GroupCache
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictUsersAdmission{})
+var _ = oadmission.WantsGroupCache(&restrictUsersAdmission{})
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+func (q *restrictUsersAdmission) SetGroupCache(c *usercache.GroupCache) {
+	q.groupCache = c
+}
+
+// objectReferenceDelta returns the relative complement of
+// []ObjectReference elementsToIgnore in []ObjectReference elements
+// (i.e., elements∖elementsToIgnore).
+func objectReferenceDelta(elementsToIgnore, elements []kapi.ObjectReference) []kapi.ObjectReference {
+	result := []kapi.ObjectReference{}
+
+	for _, el := range elements {
+		keep := true
+		for _, skipEl := range elementsToIgnore {
+			if el == skipEl {
+				keep = false
+				break
+			}
+		}
+		if keep {
+			result = append(result, el)
+		}
+	}
+
+	return result
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// each subject in the binding must be matched by some rolebinding restriction
+// in the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if !q.groupCache.Running() {
```

Comment:
move this check after check of the resource so that you don't fail requests you don't have to.